diff --git a/includes/common.inc b/includes/common.inc
index cd30145..6f26f9e 100644
--- a/includes/common.inc
+++ b/includes/common.inc
@@ -5497,6 +5497,11 @@ function drupal_system_listing($mask, $directory, $key = 'name', $min_depth = 1)
     $searchdir[] = "$config/$directory";
   }
 
+  // To allow component modules to be discovered in the theme
+  if($directory == 'modules'){
+    $searchdir[] = "profiles/$profile/themes";
+  }
+
   // Get current list of items.
   if (!function_exists('file_scan_directory')) {
     require_once DRUPAL_ROOT . '/includes/file.inc';
@@ -5529,6 +5534,15 @@ function drupal_system_listing($mask, $directory, $key = 'name', $min_depth = 1)
     $files = array_merge($files, $files_to_add);
   }
 
+  // Avoid components to be added as themes
+  if($directory == 'themes'){
+    foreach($files as $file_key => $file){
+      if(preg_match('/kostym_components\/' . $file_key . '\/' . $file_key . '\.info$/', $file->uri)){
+        unset($files[$file_key]);
+      }
+    }
+  }
+
   return $files;
 }
 
